File: android/src/main/java/com/reactnativechimesdk/views/ChimeView.java
package com.reactnativechimesdk.views;

import android.content.Context;
import android.os.Handler;
import android.util.AttributeSet;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.LifecycleOwner;

import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.DefaultVideoRenderView;
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoPauseState;
import com.annimon.stream.Stream;
import com.facebook.react.bridge.ReactContext;
import com.reactnativechimesdk.R;

import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

import static com.reactnativechimesdk.MeetingModel.meetingModel;

public class ChimeView extends FrameLayout {

  private static final String TAG = "ChimeView";

  private final Handler mainHandler = new Handler();

  private DefaultVideoRenderView renderView;
  private String attendeeId;
  private ScheduledFuture<?> scheduledFuture;

  public ChimeView(@NonNull Context context) {
    this(context, null);
  }

  public ChimeView(@NonNull Context context, @Nullable AttributeSet attrs) {
    this(context, attrs, 0);
  }

  public ChimeView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);

    LifecycleOwner owner = (AppCompatActivity) ((ReactContext) context).getCurrentActivity();
    assert owner != null;
    meetingModel().getVideoTilesLive().observe(owner, videoCollectionTiles -> bind());
  }

  @Override
  protected void onFinishInflate() {
    super.onFinishInflate();
    renderView = findViewById(R.id.videoView);
  }

  public void bindVideoTile(String attendeeId) {
    boolean viewVisible = !attendeeId.isEmpty();
    if (viewVisible) {
      this.attendeeId = attendeeId;
      bind();
    } else {
      unbind();
    }
  }

  private void bind() {
    Stream.of(meetingModel().getVideoTiles())
      .filter(it -> it.getVideoTileState().getAttendeeId().equals(ChimeView.this.attendeeId))
      .findSingle()
      .executeIfAbsent(() -> setVisibility(GONE))
      .executeIfPresent(v -> {
        scheduledFuture = meetingModel().executorService.schedule(() -> {
          try {
            if (v.getVideoTileState().getPauseState() == VideoPauseState.PausedByUserRequest) {
              meetingModel().getAudioVideo().resumeRemoteVideoTile(v.getVideoTileState().getTileId());
            }
            meetingModel().getAudioVideo().bindVideoView(renderView, v.getVideoTileState().getTileId());
          } catch (Exception e) {
            e.printStackTrace();
          }
          mainHandler.post(() -> setVisibility(VISIBLE));
        }, 200, TimeUnit.MILLISECONDS);
      });
  }

  private void unbind() {
    Stream.of(meetingModel().getVideoTiles())
      .filter(it -> it.getVideoTileState().getAttendeeId().equals(ChimeView.this.attendeeId))
      .findSingle()
      .executeIfAbsent(() -> setVisibility(GONE))
      .executeIfPresent(v -> {
        scheduledFuture = meetingModel().executorService.schedule(() -> {
          try {
            meetingModel().getAudioVideo().pauseRemoteVideoTile(v.getVideoTileState().getTileId());
          } catch (Exception e) {
            e.printStackTrace();
          }
          mainHandler.post(() -> setVisibility(GONE));
        }, 200, TimeUnit.MILLISECONDS);
      });
  }

  @Override
  protected void onDetachedFromWindow() {
    super.onDetachedFromWindow();
    if (scheduledFuture != null) {
      scheduledFuture.cancel(true);
      scheduledFuture = null;
    }
    mainHandler.removeCallbacksAndMessages(null);
  }
}

File: android/src/main/java/com/reactnativechimesdk/views/ChimeView.kt
package com.reactnativechimesdk.views

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.widget.FrameLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.DefaultVideoRenderView
import com.facebook.react.bridge.ReactContext
import com.reactnativechimesdk.ChimeVideoViewManager
import com.reactnativechimesdk.MeetingSessionModel
import com.reactnativechimesdk.R

class ChimeView : FrameLayout {

  private lateinit var renderView: DefaultVideoRenderView

  private var attendeeId: String? = null
  private var success: Boolean = false
  private var visible: Boolean = false

  constructor(context: Context) : super(context)
  constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
  constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

  init {
      MeetingSessionModel.getRemoteVideosLiveData().observe(((context as ReactContext).currentActivity as AppCompatActivity), Observer { list ->
        list.forEach {
          if (it.videoTileState.attendeeId == attendeeId && !success && visible) {
            Log.d(ChimeVideoViewManager.TAG, "bind remote video $attendeeId")
            success = true
            MeetingSessionModel.audioVideo.bindVideoView(renderView, it.videoTileState.tileId)
          }
        }
      })
  }

  override fun onFinishInflate() {
    super.onFinishInflate()
    renderView = findViewById(R.id.videoView)

  }

  fun bindVideoTile(attendeeId: String) {
    visible = attendeeId.isNotEmpty()
    if (visible) {
      this.attendeeId = attendeeId
      bind()
    } else {
      unbind()
    }
  }

  private fun bind() {
    if (isLocal()) {
      MeetingSessionModel.localVideoTileState?.videoTileState?.let {
        Log.d(ChimeVideoViewManager.TAG, "bind local video")
        success = true
        MeetingSessionModel.audioVideo.bindVideoView(renderView, it.tileId)
      }
    } else {
      MeetingSessionModel.remoteVideoTileStates.forEach {
        if (it.videoTileState.attendeeId == attendeeId) {
          Log.d(ChimeVideoViewManager.TAG, "bind remote video $attendeeId")
          success = true
          MeetingSessionModel.audioVideo.bindVideoView(renderView, it.videoTileState.tileId)
        }
      }
    }

    if (!success) {
      Log.d(ChimeVideoViewManager.TAG, "Video tile is not available on attendee: $attendeeId")
    }
  }

  private fun unbind() {
    if (isLocal()) {
      MeetingSessionModel.localVideoTileState?.videoTileState?.let {
        MeetingSessionModel.audioVideo.unbindVideoView(it.tileId)
      }
    } else {
      MeetingSessionModel.remoteVideoTileStates.forEach {
        if (it.videoTileState.attendeeId == attendeeId) {
          MeetingSessionModel.audioVideo.unbindVideoView(it.videoTileState.tileId)
        }
      }
    }
  }

  private fun isLocal(): Boolean {
    return MeetingSessionModel.localAttendeeId == attendeeId
  }
}
